internal/database/testing/database.go

```diff
@@ -14,102 +14,138 @@ import (
+	for _, table := range tables {
+		dumpTable(c, db, table, newJSONFormatter(c))
 	}
-	fmt.Fprintln(writer)
+}
 
-	vals := make([]any, len(cols))
-	for i := range vals {
-		vals[i] = new(any)
+func dumpTable(c *gc.C, db *sql.DB, tableName string, f formatter) {
+	f.SetTableName(tableName)
+	rows, err := db.Query("SELECT * FROM " + tableName)
+	if err != nil {
+		// Soft fail, most likely the table doesn't exist
+		fmt.Println(err)
+		return
 	}
 
-	for rows.Next() {
-		err = rows.Scan(vals...)
-		c.Assert(err, jc.ErrorIsNil)
+	columns, err := rows.Columns()
+	c.Assert(err, jc.ErrorIsNil)
+	f.SetColumns(columns)
 
-		for _, val := range vals {
-			fmt.Fprintf(writer, "%v\t", *val.(*any))
+	for rows.Next() {
+		row := make([]any, len(columns))
+		for i := range row {
+			row[i] = new(any)
 		}
+		err = rows.Scan(row...)
+		c.Assert(err, jc.ErrorIsNil)
+		f.AddRow(row)
 	}
-	err = rows.Err()
-	c.Assert(err, jc.ErrorIsNil)
+	f.Flush()
+}
 
-	writer.Flush()
-	fmt.Fprintln(os.Stdout)
+// formatter defines a way of printing a table.
+type formatter interface {
+	// SetTableName defines the name of the table
+	SetTableName(string)
+	// SetColumns defines the columns of the table.
+	SetColumns([]string)
+	// AddRow adds a row to the table
+	AddRow([]any)
+	// Flush prints the table.
+	Flush()
 }
 
-// DumpTablesJSON prints the named tables to stdout in a JSON format. This is
-// intended as a debugging tool for state tests. It is not intended for use in
-// production code.
-func DumpTablesJSON(c *gc.C, db *sql.DB, tableNames ...string) {
-	for _, tableName := range tableNames {
-		query := fmt.Sprintf("SELECT * FROM %s", tableName)
-		table := &tableData{
-			tableName: tableName,
-		}
+// tabularFormatter formats a table in a tabular format.
+type tabularFormatter struct {
+	tableName string
+	tw        *tabwriter.Writer
+}
 
-		rows, err := db.Query(query)
-		if err != nil {
-			// Soft fail, most likely the table doesn't exist
-			fmt.Println(err)
-			continue
-		}
+func newTabularFormatter() *tabularFormatter {
+	return &tabularFormatter{
+		tw: tabwriter.NewWriter(os.Stdout, 0, 8, 1, '\t', 0),
+	}
+}
 
-		columns, err := rows.Columns()
-		c.Assert(err, jc.ErrorIsNil)
-		table.columnNames = columns
-
-		for rows.Next() {
-			row := make([]any, len(columns))
-			for i := range row {
-				row[i] = new(any)
-			}
-			err = rows.Scan(row...)
-			c.Assert(err, jc.ErrorIsNil)
-			table.rows = append(table.rows, row)
-		}
+func (t *tabularFormatter) SetTableName(name string) {
+	t.tableName = name
+}
 
-		printJSON(c, table)
+func (t *tabularFormatter) SetColumns(cols []string) {
+	for _, col := range cols {
+		fmt.Fprintf(t.tw, "%s\t", col)
 	}
+	fmt.Fprintln(t.tw)
+}
 
+func (t *tabularFormatter) AddRow(vals []any) {
+	for _, val := range vals {
+		fmt.Fprintf(t.tw, "%v\t", *val.(*any))
+	}
+}
+
+func (t *tabularFormatter) Flush() {
+	fmt.Fprintln(os.Stdout)
+	fmt.Printf("***** TABLE %q *****\n", t.tableName)
```

Comment:
```suggestion
	fmt.Fprintf(os.Stdout, "Table: %q\n", t.tableName)
	fmt.Fprintf(os.Stdout, strings.Repeat("-", len(t.tableName) + 7))
```

You don't need to scream about it.